kernel/src/log_compaction/writer.rs

```diff
@@ -0,0 +1,265 @@
+use crate::checkpoint::log_replay::CheckpointLogReplayProcessor;
+use crate::log_replay::LogReplayProcessor;
+
+/// Utility function to determine if log compaction should be performed
+/// based on the commit version and compaction interval.
+#[allow(dead_code)]
+pub(crate) fn should_compact(commit_version: Version, compaction_interval: Version) -> bool {
+    // Commits start at 0, so we add one to the commit version to check if we've hit the interval
+    compaction_interval > 0
+        && commit_version > 0
+        && ((commit_version + 1) % compaction_interval) == 0
+}
+
+/// Writer for log compaction files
+///
+/// This writer provides an API similar to [`CheckpointWriter`] for creating
+/// log compaction files that aggregate actions from multiple commit files.
+///
+/// [`CheckpointWriter`]: crate::checkpoint::CheckpointWriter
+#[derive(Debug)]
+#[allow(dead_code)]
+pub(crate) struct LogCompactionWriter {
+    table_root: Url,
+    start_version: Version,
+    end_version: Version,
+    compaction_path: ParsedLogPath<Url>,
+    /// Minimum file retention timestamp for tombstone cleanup
+    min_file_retention_timestamp_millis: i64,
+}
+
+impl LogCompactionWriter {
+    #[allow(dead_code)]
+    pub(crate) fn new(
+        table_root: Url,
+        start_version: Version,
+        end_version: Version,
+        min_file_retention_timestamp_millis: i64,
+    ) -> DeltaResult<Self> {
+        if start_version > end_version {
+            return Err(Error::generic(format!(
+                "Invalid version range: start_version {} > end_version {}",
+                start_version, end_version
+            )));
+        }
+
+        let compaction_path =
+            ParsedLogPath::new_log_compaction(&table_root, start_version, end_version)?;
+
+        Ok(Self {
+            table_root,
+            start_version,
+            end_version,
+            compaction_path,
+            min_file_retention_timestamp_millis,
+        })
+    }
+
+    /// Get the path where the compaction file will be written
+    #[allow(dead_code)]
+    pub(crate) fn compaction_path(&self) -> DeltaResult<Url> {
+        Ok(self.compaction_path.location.clone())
+    }
+
+    /// Get an iterator over the compaction data to be written
+    ///
+    /// This method performs action reconciliation similar to checkpoint creation,
+    /// but specifically for the version range specified in the constructor.
+    /// It reuses the CheckpointLogReplayProcessor to ensure consistent reconciliation
+    /// logic with checkpoint creation.
+    #[allow(dead_code)]
+    pub(crate) fn compaction_data(
+        &mut self,
+        engine: &dyn Engine,
+    ) -> DeltaResult<LogCompactionDataIterator> {
+        // List commit files in the specified range
+        let commit_files = self.list_commit_files(engine)?;
+
+        // Validate that we have the expected number of commit files
+        let expected_count = (self.end_version - self.start_version + 1) as usize;
```

Comment:
`as` is generally considered unsafe (lossy) - in this instance it will silently truncate if `usize` > `u64` or will wrap negatives to huge `usize`. (generally we stay away from `as` since there are ways perform fallible conversions)

here we could use `checked_sub`/`checked_add` or `saturating_*` and just use `try_into` for conversion